### ISSUE TITLE
Can this be used on android?

### ISSUE THREAD

**Issue by bj0** [kunalsheth/units-of-measure#7]
This looks like a very fascinating project, and I would like to try it out in an android app that I have to do a lot of unit conversion in.  The configuration on the wiki, unfortunately, does not work for android.  

I get
>Could not get unknown property 'sourceSets' for root project

Is it possible to adjust the config to work on android?

**Comment by kunalsheth** [comment]
Hi Brian,
I haven't gotten a chance to update the wiki recently (though I am, and will continue to, maintain the library itself).

I recommend creating a separate module just for _units-of-measure_ and then depending on said module from your project. This setup will also significantly shorten build times.

**Comment by kunalsheth** [comment]
Here is an example project I just created: https://github.com/kunalsheth/uom-example-project
the _uoms_ module contains all configuration for _units-of-measure_. _myapp_ depends on _uoms_.

Units-of-measure is being used in production on my robotics team, @Team846. I've copied and pasted some of our robot control utilities and unit tests into _myapp_. Hopefully, the same Gradle setup works on android as well.

Let me know how well this works for you.
-Kunal

**Comment by kunalsheth** [comment]
In addition to the functionality shown in the sample code. I've also built a type-safe wrapper over kotlin.math. This makes _units-of-measure_ a lot more powerful because it can verify more complex equations.

[Here](https://github.com/kunalsheth/uom-example-project/blob/9e7b9ede2e16baa1ed2eda582f992cc0a9b61d38/myapp/src/main/kotlin/com/lynbrookrobotics/kapuchin/control/math/kinematics/Pathing.kt#L40) is an example where _units-of-measure_ verifies that I used the law of cosines correctly.

**Comment by bj0** [comment]
So an android project has a different structure, and there was no `kotlinCompile` task,  I was able to get it to work by removing the lines at the bottom of the `units-of-measure.gradle` file, and just using the following `build.gradle` for a java library module in android:

```
apply plugin: 'kotlin'
apply plugin: 'java-library'
apply plugin: 'info.kunalsheth.units.gradle'

dependencies {
    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
}

apply from: 'units-of-measure.gradle'

tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).all {
    dependsOn generateUnitsOfMeasure
}

sourceSets.main.java.srcDirs += generateUnitsOfMeasure.generatedSrcDir

sourceCompatibility = "8"
targetCompatibility = "8"
```

**Comment by bj0** [comment]
Gotta be honest though, I've read the wiki, the configs, and the generated code a couple times now and I still am not sure how the `relation` thing in the config relates to real uses.

**Comment by kunalsheth** [comment]
Relation is a way of telling the plugin which calculations you want to be able to do.
Every unit is of a dimension. (E.g. unit _km/h_ is of dimension _velocity_). Each dimension made up of base dimensions.

The problem is that there are infinite combinations of base dimensions possible. While certain, more academic languages (E.g. [Scala](https://typelevel.org/blog/2017/06/13/libra.html)) allow users to represent infinite types. Kotlin requires that we actually write out (or in this case, generate) a new class for every type we want to use.

tl;dr you have to tell _units-of-measure_ which operations/units/dimensions you want to use so it can actually generate them.

Let's say I want to use _units-of-measure_ for some electrical calculations. In that case, my `relationships` would look something like this: 
```gradle
relationships += r(
            d(M: 1, L: 2, T: -3, I: -1), // volts
            d(M: 1, L: 2, T: -3, I: -2), // ohms
            d(I: 1), // amps
            d(T: 1), // time
    )
```

This tells _units-of-measure_ to generate operations involving volts, amps, ohms, and time.
It will basically generate the code needed to multiply and divide any combination of said electrical units.

**Comment by kunalsheth** [comment]
If you tell me a little about what type of calculations you generally plan on doing, I can help you write your configuration file.

**Comment by bj0** [comment]
Ok, I think I understand that now.  For my use case I'm only converting between different types of units (fps to m/s, etc), not doing any calculations with them.  I think the only relations I need then are between L and T, as one of my units is velocity.

EDIT: this is what I have so far:
```
generateUnitsOfMeasure {
    relationships += r(d(L: 1), d(T: 1))

    quantities += [
//            q("BallisticCoefficient", d(M: 1, L: -2)),
    ].toSet()

    unitsOfMeasure += [
            u("Grain", 6.47989e-5, d(M: 1)),
//            u("Gram", 1e-3, d(M: 1)),
            u("Ounce", 0.0283495, d(M: 1)),
            u("Pound", 0.453592, d(M: 1)),

            u("Degree", Math.PI / 180, d()),

            u("Percent", 0.01, d()),

            u("Inch", 0.0254, d(L: 1)),
            u("Foot", 0.3048, d(L: 1)),
            u("Yard", 0.9144, d(L: 1)),
            u("Mile", 1609.344, d(L: 1)),

            u("Minute", 60.0, d(T: 1)),
            u("Hour", 3600.0, d(T: 1)),

            u("Mps", 1.0, d(L: 1, T: -1)),
            u("Fps", 0.3048, d(L: 1, T: -1)),
            u("mph", 0.44704, d(L: 1, T: -1)),
            u("Knots", 0.514444, d(L: 1, T: -1)),

            u("Bar", 1e5, d(M: 1, L: -1, T: -2)),
            u("InchOfHg", 3386.39, d(M: 1, L: -1, T: -2)),
            u("Atm", 101325, d(M: 1, L: -1, T: -2)),
            u("Psi", 6894.76, d(M: 1, L: -1, T: -2)),
    ].toSet()
}

```

EDIT2: Updated for newer version.

Is there a way to add Celsius and Fahrenheit?

**Comment by kunalsheth** [comment]
Celsius and Fahrenheit would have to be manually programmed in on the user's end because Celsius and Fahrenheit aren't "units" in the theoretical sense because they have different zeros.

To add Celsius and Fahrenheit, your code would look like the following:
```kotlin
inline val Number.Celsius get() = (toDouble() + 273.15).Kelvin
inline val `Θ`.Celsius get() = siValue - 273.15
object `Celsius` : UomConverter<`Θ`>,
    Quan<`Θ`> by box(1.Celsius) {
    override val unitName = "Celsius"
    override fun invoke(x: Double) = x.Celsius
    override fun invoke(x: `Θ`) = x.Celsius
}

inline val Number.Fahrenheit get() = ((toDouble() − 32) * 5/9).Celsius
inline val `Θ`. Fahrenheit get() = (siValue − 273.15) * 9/5 + 32
object Fahrenheit : UomConverter<`Θ`>,
    Quan<`Θ`> by box(1.Fahrenheit) {
    override val unitName = "Fahrenheit"
    override fun invoke(x: Double) = x.Fahrenheit
    override fun invoke(x: `Θ`) = x.Fahrenheit
}
```

note: I haven't testing this. Please double-check my conversion math for the above code.

**Comment by bj0** [comment]
Ok it looks like everything is working well, thank you.  I really appreciate this library.  

I do have one more question, though:  How do you deal with rounding errors and comparing values. 

For example, I have data classes with `Length` properties, and when I compare two of them after some conversions I get something like `0.03809999942779541 m` vs `0.038099999999999995 m`.  Have you run into this issue, and if so do you do something special to compare them?

**Comment by kunalsheth** [comment]
Hi Brian, I definitely have run into this issue.
When checking if two values are equal, I use a tolerance, either with Kotlin's `..` range syntax or with _units-of-measure_'s `±` functions. If you don't want to create an instance of `ClosedRange` for every comparison, then you can use an expression like `(a - b).abs < tolerance` instead.

For examples on how this is done, take a look at the [sample code](https://github.com/kunalsheth/units-of-measure/blob/master/demo/src/main/kotlin/info/kunalsheth/units/sample/Sample.kt#L13).

Additionally, my robotics team has written a lot of unit tests using _units-of-measure_. You can take a look at some of the utility functions we wrote [here](https://github.com/kunalsheth/uom-example-project/blob/master/myapp/src/test/kotlin/com/lynbrookrobotics/kapuchin/tests/Utils.kt).

**Comment by bj0** [comment]
Ah those are pretty useful.  I was trying to just compare 2 data classes full of values, but I guess it just can't be that easy...

Thanks.

**Comment by kunalsheth** [comment]
Ahh, that is unfortunate. I'll have to think of a workaround.

Currently, Kotlin does not allow you to override `equals` in an inline class, so I'm not sure how to go about adding in default tolerances when comparing quantities.

That being said, you would eventually encounter this very same issue with doubles as well. _units-of-measure_ just makes it more apparent because there are a couple more `*` and `/`s taking place under the hood.

**Comment by kunalsheth** [comment]
A potential solution may be to truncate/round all the quantities in your data class before calling the `equals` function. This is less verbose than rewriting the auto-generated `equals` method using ranges and you'll also get some cleaner logging.

```kotlin
/**
 * Rounds a number to a fixed number of decimal places for cleaner logging
 *
 * @author Kunal
 *
 * @receiver number to round
 * @param decimalPlaces number of digits to keep left of the decimal. Negative numbers round to the right of the decimal.
 * @return a rounded double
 */
infix fun Number.withDecimals(decimalPlaces: Int) = toDouble().let {
    val shifter = 10.0.pow(decimalPlaces)
    round(it * shifter) / shifter
}

/**
 * Rounds a number to a fixed number of decimal places for cleaner logging
 *
 * @author Kunal
 *
 * @receiver number to round
 * @param decimalPlaces number of digits to keep left of the decimal. Negative numbers round to the right of the decimal.
 * @return a rounded quantity
 */
infix fun <Q : Quan<Q>> Q.withDecimals(decimalPlaces: Int) = new(siValue withDecimals decimalPlaces)
```

**Comment by bj0** [comment]
fyi I just upgraded to 5.3.0 and it no longer works as a normal jvm library.  It gives this error:

> UomMath.kt: (5, 18): Declaration annotated with '@OptionalExpectation' can only be used in common module sources

I converted it to a multiplatform module and changed the directory structure to try and match common sources (I don't really have any experience with mp projects).  It does compile and seem to work now, but the source files no longer show up in Android Studio, and the tests I had included in the `uom` project stopped running.

Do you do run jvm tests in your uom module?

edit: looks like it's the use of `JsName`

**Comment by kunalsheth** [comment]
Hi Brian,
Sorry I broke your code. I'm aware of this issue.
I've never run JVM tests in my UOM module because I didn't intend the UOM module to contain user code. Maybe you can put your code in a separate module which depends on UOM?

Regarding the source files not showing up and tests not running, I'm not sure, but it sounds a bit like the module wasn't imported in `settings.gradle`? 

I've recently published a website about the library, you can see the suggested module configuration here: http://units.kunalsheth.info/setup/install.html

**Comment by bj0** [comment]
I actually got my new config from your website.  I just added the following source set:
```        jvmTest {
            dependencies {
                implementation kotlin('test')
                implementation kotlin('test-junit')
                implementation "com.google.truth:truth:0.45"
            }
        }
```

The project is included in `settings.gradle`, the modules show up in the project view on the left, they just have no items underneath them (except `uom_jvmTest` now shows the unit tests).  

I could move the unit tests to another module, I only put them in there because they only deal with code inside the `uom` module.

**Comment by kunalsheth** [comment]
Oh, I see. Give me a moment, I'll try to replicate this.

**Comment by kunalsheth** [comment]
I believe I replicated the problem. Modules appear, but tests don't run. Working on a fix.

**Comment by kunalsheth** [comment]
Hi Brian, it seems to just be a configuration thing with MP testing.
See: https://kotlinlang.org/docs/reference/building-mpp-with-gradle.html#running-tests

Here's what I changed to get my test to run:
```diff
diff --git a/uoms/build.gradle b/uoms/build.gradle
index 1eb2270..ae49306 100644
--- a/uoms/build.gradle
+++ b/uoms/build.gradle
@@ -14,6 +14,13 @@ kotlin {
             dependencies { implementation kotlin('stdlib-common') }
         }
         jvm { dependencies { implementation kotlin('stdlib-jdk8') } }
+        jvmTest {
+            dependencies {
+                implementation kotlin('test')
+                implementation kotlin('test-junit')
+                implementation "com.google.truth:truth:0.45"
+            }
+        }
         js { dependencies { implementation kotlin('stdlib-js') } }
         all { languageSettings.enableLanguageFeature('InlineClasses') }
     }
diff --git a/uoms/src/jvmTest/kotlin/MyTestTest.kt b/uoms/src/jvmTest/kotlin/MyTestTest.kt
new file mode 100644
index 0000000..5d1f158
--- /dev/null
+++ b/uoms/src/jvmTest/kotlin/MyTestTest.kt
@@ -0,0 +1,15 @@
+package replicate.bug.test
+
+import org.junit.Test
+
+class MyTestTest {
+    @Test
+    fun thisShouldPass() {
+        assert(1 + 1 == 2)
+    }
+
+    @Test
+    fun thisShouldFail() {
+        assert(2 + 2 == 5)
+    }
+}
\ No newline at end of file
```

**Comment by kunalsheth** [comment]
That being said, may I ask what you are trying to test inside the UOM module?
Are you concerned that the generated code may be incorrect?

**Comment by bj0** [comment]
I guess I was just doing some conversion tests to verify the units I am using are converting to what I expect them to (within tolerances).  I did put that section in, and I can get it to run the tests in gradle, but it doesn't display the output (it generates a report xml).  I'm not too familiar with using gradle to run tests, though.

I also noticed that you have `jvm {` defined, but I think it's supposed to be `jvmMain {`.  This is the config I have now:
```apply plugin: 'org.jetbrains.kotlin.multiplatform'
apply plugin: 'info.kunalsheth.units.gradle'

repositories.mavenCentral()

kotlin {
    jvm()

    sourceSets {
        commonMain {
            kotlin.srcDirs += generateUnitsOfMeasure.generatedSrcDir
            dependencies { implementation kotlin('stdlib-common') }
        }
        jvmMain {
            dependsOn commonMain
            dependencies {
                implementation kotlin('stdlib')
            }
        }
        jvmTest {
            dependencies {
                implementation kotlin('test')
                implementation kotlin('test-junit')
                implementation "com.google.truth:truth:0.45"
            }
        }
        all {
            languageSettings.enableLanguageFeature('InlineClasses')
        }
    }
    targets.all {
        compilations.all { compileKotlinTask.dependsOn(generateUnitsOfMeasure) }
    }
}
apply from: 'units-of-measure.gradle'

```

It works, but AS still shows empty modules on the project (Android) display and AS can't run the tests (but gradle can).  This might just be Android Studio's limited support for multiplatform projects though.

**Comment by kunalsheth** [comment]
Hmm.. I think the reason AS doesn't run them is because `:uoms` doesn't have a `test` task.
Instead, Kotlin MP has `allTests` and `check`.

When I run `:check` in Intellij, my unit test runs and stdout and stderr are displayed.

**Comment by bj0** [comment]
Yea, when I run from the commandline it will *run* the tests, but it doesn't print any output (I have to look up the .xml report).  Hopefully they can make it work better with AS.  AS still doesn't show any source files in the Project Pane (in Android view).  These probably are issues with AS and not the library though.
